### ISSUE TITLE
MySQL Discovery Honors Foreign Keys Made To The Same Column Only Once

### ISSUE THREAD

**Issue by shaheero** [strongloop/loopback-connector-mysql#253]
### Bug or feature request

- [x] Bug
- [ ] Feature request

### Description of feature (or steps to reproduce if bug)
I have 2 columns that reference the same column in another table through a foreign key. MySQL discovery only made one relationship and ignored the other.

**Comment by mclark-newvistas** [comment]
Ran into something essentially similar from my import script - I was naming the relationship for the table name linked to and using that as an Object key. Only the last relation to the same table was preserved, as it overwrote the others. I ended making manual changes to the generated model files to add those and hasOne/hasMany relations, since the one vs many seems also difficult to discover from the database structure.

**Comment by ssh24** [comment]
Closing this issue due to lack of inactivity.

**Comment by shaheero** [comment]
Then I suggest you join the conversation to make it active :/

**Comment by ssh24** [comment]
There has been no description of the bug. (i.e reproduction of the bug) Feel free to open the issue if you could reproduce the issue with clear steps and example code.

**Comment by shaheero** [comment]
2 columns referencing the same FK column in MySQL, MySQL discovery only made one relationship and ignored the other. What's there to describe more?

**Comment by ssh24** [comment]
Please follow on how to report an issue: http://loopback.io/doc/en/contrib/Reporting-issues.html

PS. The ability to provide `example code` for reproduction of a bug is important for us to determine if it is indeed a bug.

**Comment by shaheero** [comment]
MySQL:
```
CREATE DATABASE `sample`;

DROP TABLE IF EXISTS `info`;

CREATE TABLE `info` (
  `id` int(11) unsigned NOT NULL AUTO_INCREMENT,
  `dept` int(11) unsigned DEFAULT NULL,
  `dept2` int(11) unsigned DEFAULT NULL,
  PRIMARY KEY (`id`),
  KEY `fk1` (`dept`),
  KEY `fk2` (`dept2`),
  CONSTRAINT `fk1` FOREIGN KEY (`dept`) REFERENCES `dept` (`id`) ON UPDATE CASCADE,
  CONSTRAINT `fk2` FOREIGN KEY (`dept2`) REFERENCES `dept` (`id`) ON UPDATE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=utf8;

DROP TABLE IF EXISTS `dept`;

CREATE TABLE `dept` (
  `id` int(11) unsigned NOT NULL AUTO_INCREMENT,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8;

```

After running a discovery script, here is the Info.json that results. It only honors one FK relationship instead of all of the ones in the info table.

```
{
  "name": "Info",
  "options": {
    "idInjection": false,
    "mysql": {
      "schema": "delete",
      "table": "info"
    },
    "relations": {
      "dept": {
        "model": "Dept",
        "type": "belongsTo",
        "foreignKey": "dept2"
      }
    }
  },
  "properties": {
    "id": {
      "type": "Number",
      "required": true,
      "length": null,
      "precision": 10,
      "scale": 0,
      "id": 1,
      "mysql": {
        "columnName": "id",
        "dataType": "int",
        "dataLength": null,
        "dataPrecision": 10,
        "dataScale": 0,
        "nullable": "N"
      }
    },
    "dept": {
      "type": "Number",
      "required": false,
      "length": null,
      "precision": 10,
      "scale": 0,
      "mysql": {
        "columnName": "dept",
        "dataType": "int",
        "dataLength": null,
        "dataPrecision": 10,
        "dataScale": 0,
        "nullable": "Y"
      }
    },
    "dept2": {
      "type": "Number",
      "required": false,
      "length": null,
      "precision": 10,
      "scale": 0,
      "mysql": {
        "columnName": "dept2",
        "dataType": "int",
        "dataLength": null,
        "dataPrecision": 10,
        "dataScale": 0,
        "nullable": "Y"
      }
    }
  }
}
```

**Comment by ssh24** [comment]
@shaheero Thank you for the example reproduction. Will reopen the issue and take a better look at it.

**Comment by ssh24** [comment]
@shaheero I just tried a reproduction of this issue and talked with relation experts /cc @jannyHou. 

Upon discovering the table schema from mysql, I would not get back any relation. That is because ATM we do not support discovering table relationships and converting them into model relationships. This feature is currently on our roadmap.

However, my question to you is, how were you able to get the `relations` object back when we do not support it? Here is my script followed by the output:
```
'use strict';

var loopback = require('loopback');

var ds = loopback.createDataSource('mysql', {
  'host': 'localhost',
  'port': 3306,
  'database': 'sample',
  'username': <my_user>,
  'password': <my_pass>,
});

// Create a model definition by discovering the given table
ds.discoverSchema('info', function(err, table) {
  if (err) throw err;
  console.log(table);
});
```
**Output**:
```
bash-3.2$ node server/bin/script.js
{ name: 'Info',
  options:
   { idInjection: false,
bash-3.2$ node server/bin/script.js
{ name: 'Info',
  options:
   { idInjection: false,
     mysql: { schema: 'sample', table: 'info' } },
  properties:
   { id:
      { type: 'Number',
        required: true,
        length: null,
        precision: 10,
        scale: 0,
        id: 1,
        mysql: [Object] },
     dept:
      { type: 'Number',
        required: false,
        length: null,
        precision: 10,
        scale: 0,
        mysql: [Object] },
     dept2:
      { type: 'Number',
        required: false,
        length: null,
        precision: 10,
        scale: 0,
        mysql: [Object] } } }
```

**Comment by shaheero** [comment]
In general, set the schema to discover as the database (not just one table) and add this attribute:
'''associations: true'''

I can provide the code in few hours if you want

**Comment by ssh24** [comment]
Not sure what you mean. You can only discover the schema of a table not the database.
For example, this code will give an error:
```
// Create a model definition by discovering the given table
ds.discoverSchema('sample', function(err, database) {
  if (err) throw err;
  console.log(database);
});
```
**Error**: `Error: Table 'sample' does not exist.`

**Comment by mclark-newvistas** [comment]
https://apidocs.strongloop.com/loopback-datasource-juggler/#datasource-prototype-discoverschemas - note this is `discoverSchemas` instead of your call to `discoverSchema`, which the documentation explicitly states "without following the relations". Note the `relations` option to `discoverSchemas`. Replication from here should be obvious.

**Comment by ssh24** [comment]
@shaheero @mclark-newvistas I was able to reproduce the bug and submit a fix for it. Now after discovering the schemas, it returns both the foreign keys as expected.

Since it is not released yet, you would need to install the dependencies from the local registry. On your app folder, do the following and then run the sample script:
```
bash-3.2$ rm -rf node_modules
bash-3.2$ npm i --registry http://ci.strongloop.com:4873/
```

**Sample script**:
```
'use strict';
var loopback = require('loopback');
var ds = loopback.createDataSource('mysql', {
  'host': 'localhost',
  'port': 3306,
  'database': 'sample',
  'username': '<my_user>',
  'password': '<my_pass>',
});

// Create a model definition by discovering the given table
ds.discoverSchemas('info', {schema: 'sample', relations: true},
function(err, table) {
  if (err) throw err;
  console.log(JSON.stringify(table['sample.info'].options.relations));
});
```
**Output**:
```
bash-3.2$ node server/bin/script.js
{"fk1rel":{"model":"Dept","type":"belongsTo","foreignKey":"dept"},"fk2rel":{"model":"Dept","type":"belongsTo","foreignKey":"dept2"}}
```

Please let me know if this fixed the original issue.

**Comment by ssh24** [comment]
Closing this as it has been fixed on the latest released juggler `v3.5.0`.
